### PR TITLE
emit warning if runtime and app mode explicitly disagree

### LINKED ISSUE #29350: Open Telmeetry needs to emit a warning if in app mode
As per the UFO we must emit warnings when

- otel.sdk.disabled=true is set with application configuration but otel.sdk.disabled=false is set by the server configuration.
- MicroProfile Telemetry is configured by the application instead of the server. Consequently, runtime metrics and logs are disabled.

### PR DESCRIPTION
**Pull request by benjamin-confino**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Resolves #29350

TODO
- [x] Review messages
- [x] FAT tests.

### CODE REVIEW

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 1]
```diff
@@ -55,6 +55,10 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
 CWMOT5005.mptelemetry.unknown.log.source.explanation=The Liberty runtime environment does not provide the specified log source.
 CWMOT5005.mptelemetry.unknown.log.source.useraction=Check the configuration information and specify one of the supported sources for MicroProfile Telemetry.
 
+CWMOT5006.tel.enabled.conflict=CWMOT5006W: Conflicting configuration for otel.sdk.disabled detected.
+CWMOT5006.tel.enabled.conflict.explanation: When MicroProfile Telemetry is configured to run with a shared instance of OpenTelemetry for all applications it ignores the server.xml file as a configuration source. We have detected that the value of otel.sdk.disabled in server.xml conflicts with the value set as a system property or environment variable. (System properties take precedence).
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Change to:
"When MicroProfile Telemetry is configured to run with a shared instance of OpenTelemetry for all applications, it ignores the server.xml file as a configuration source. The value of otel.sdk.disabled in the server.xml file conflicts with the value set as a system property or environment variable. System properties take precedence."

- Add comma to first sentence
- Remove "We have detected that"
- Remove parens

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 2]
```diff
@@ -55,6 +55,10 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
 CWMOT5005.mptelemetry.unknown.log.source.explanation=The Liberty runtime environment does not provide the specified log source.
 CWMOT5005.mptelemetry.unknown.log.source.useraction=Check the configuration information and specify one of the supported sources for MicroProfile Telemetry.
 
+CWMOT5006.tel.enabled.conflict=CWMOT5006W: Conflicting configuration for otel.sdk.disabled detected.
+CWMOT5006.tel.enabled.conflict.explanation: When MicroProfile Telemetry is configured to run with a shared instance of OpenTelemetry for all applications it ignores the server.xml file as a configuration source. We have detected that the value of otel.sdk.disabled in server.xml conflicts with the value set as a system property or environment variable. (System properties take precedence).     
+CWMOT5006.tel.enabled.conflict.useraction: Ensure that your server.xml, system properties and environment variables are brought into alignment.
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Add a comma after "properties" and simplify "brought into alignment"; for example, change to:

"Ensure that your server.xml, system properties, and environment variables are consistent."

**Review by helyarp — COMMENTED** [review]
Suggested changes

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops (hunk 2 above)
Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: I'm not sure about this one. With a comma after system properties that says all three of server.xml, system properies, and environment variables must be consistent. 

But what I want to say is that server.xml must be consistent with the combined [system properties + environment variables]. I will slack you to discuss.

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/publish/servers/Telemetry10IgnoreMPConfig/server.xml [hunk 3]
```diff
@@ -8,6 +8,9 @@
         <feature>mpTelemetry-1.0</feature>
         <feature>componentTest-2.0</feature>
     </featureManager>
+    
+    <!-- ignored since we will run in runtime mode but this will cause a warning -->
+	<variable name="otel.sdk.disabled" value="true"/>
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/publish/servers/Telemetry10IgnoreMPConfig/server.xml]: bad indentation

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 4]
```diff
@@ -55,6 +55,10 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
 CWMOT5005.mptelemetry.unknown.log.source.explanation=The Liberty runtime environment does not provide the specified log source.
 CWMOT5005.mptelemetry.unknown.log.source.useraction=Check the configuration information and specify one of the supported sources for MicroProfile Telemetry.
 
+CWMOT5006.tel.enabled.conflict=CWMOT5006W: Conflicting configuration for otel.sdk.disabled detected.
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: I'm not convinced this error message has enough information. If possible we should at least tell the user where the conflicting configurations were found.

**Review by Yas-Franco — DISMISSED** [review]

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/FATSuite.java [hunk 5]
```diff
@@ -22,34 +22,11 @@
...
-                TelemetryMisconfigTest.class,
-                TelemetryLongRunningTest.class,
                 TelemetryGlobalOpenTelemetryTest.class,
-                TelemetryDisabledTest.class,
-                TelemetryServletTest.class,
-                TelemetryWithSpanErrorTest.class,
-                TelemetryAttributesTest.class,
-                TelemetryRuntimeInstanceTest.class,
 })
 public class FATSuite {
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/FATSuite.java]: This change should not be committed.

File: dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java [hunk 6]
```diff
@@ -86,6 +86,10 @@ private void checkThenSetRuntimeFields(boolean telemetry2OrLater) {
         HashMap<String, String> propreties = OpenTelemetryPropertiesReader.getRuntimeInstanceTelemetryProperties();
         isRuntimeEnabled = telemetry2OrLater && !!!OpenTelemetryPropertiesReader.checkDisabled(propreties);
 
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "Configured runtime mode as ", isRuntimeEnabled);
+        }
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java]:
```suggestion
        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
            Tr.debug(this, tc, "Configured runtime mode as ", isRuntimeEnabled);
        }
```

Needed put the correct value in the `id=xxxxxxxx` field in the trace log. All trace calls in non-static methods should pass `this` as the first argument.

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 7]
```diff
@@ -55,6 +55,15 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
 CWMOT5005.mptelemetry.unknown.log.source.explanation=The Liberty runtime environment does not provide the specified log source.
 CWMOT5005.mptelemetry.unknown.log.source.useraction=Check the configuration information and specify one of the supported sources for MicroProfile Telemetry.
 
+CWMOT5006.tel.enabled.conflict=CWMOT5006W: Conflicting configuration for the otel.sdk.disabled configuration property detected for the {0} application. The final value is otel.sdk.disabled=false. Telemetry cannot be disabled for an application when it is enabled for the runtime.
+CWMOT5006.tel.enabled.conflict.explanation: MicroProfile Telemetry can be enabled for the whole server using system properties or environment variables. If it isn't enabled for the whole server it can be enabled for each application by setting MicroProfile Config properties, for example by setting variables in the server.xml file. This warning is given if there's explicit configuration disabling Telemetry for the whole server which is overridden by an MP Config property enabling Telemetry for an application.
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Pls change to:
MicroProfile Telemetry can be enabled for the whole server by using system properties or environment variables. If it isn't enabled for the whole server, it can be enabled for an application by setting MicroProfile Config properties; for example, by setting variables in the server.xml file. This warning is given if there's an explicit configuration that disables Telemetry for the whole server, which is overridden by an MP Config property that enables Telemetry for an application.
(discussed on Slack)

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 8]
```diff
@@ -55,6 +55,15 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
...
+CWMOT5006.tel.enabled.conflict.useraction: Specify the settings to enable or disable OpenTelemetry instances using either environment variables and system properties or other MP Config config sources but not both.
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Change "using" to "by using"

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 9]
```diff
@@ -55,6 +55,15 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
...
+
+CWMOT5007.tel.enabled.conflict=CWMOT5007W: Conflicting configuration for the otel.sdk.disabled configuration property detected for the {0} application. The final value is otel.sdk.disabled=false because the property enabling telemetry for the application overrides the property disabling telemetry for the runtime.
+CWMOT5007.tel.enabled.conflict.explanation: MicroProfile Telemetry can be enabled for the whole server using system properties or environment variables. If it isn't enabled for the whole server it can be enabled for each application by setting MicroProfile Config properties, for example by setting variables in the server.xml file. This warning is given if Telemetry is enabled for the whole server, but the MicroProfile Config properties for an application would indicate that Telemetry should be disabled for an application. Since Telemetry is enabled for the whole server, the configuration for that application is ignored.
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Change to:
MicroProfile Telemetry can be enabled for the whole server by using system properties or environment variables. If it isn't enabled for the whole server, it can be enabled for an application by setting MicroProfile Config properties; for example, by setting variables in the server.xml file. This warning is given if Telemetry is enabled for the whole server, but the MicroProfile Config properties disable Telemetry for an application. Because Telemetry is enabled for the whole server, the configuration for that application is ignored.

(discussed on Slack)

File: dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops [hunk 10]
```diff
@@ -55,6 +55,15 @@ CWMOT5005.mptelemetry.unknown.log.source=CWMOT5005W: The MicroProfile Telemetry
...
+CWMOT5007.tel.enabled.conflict.useraction: Specify the settings to enable or disable OpenTelemetry instances using either environment variables and system properties or other MP Config config sources but not both.
```
Comment (helyarp) [inline dev/io.openliberty.microprofile.telemetry.internal.common/resources/io/openliberty/microprofile/telemetry/internal/common/resources/MPTelemetry.nlsprops]: Change "using" to "by using"

**Review by helyarp — COMMENTED** [review]
Suggested changes

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryConfigServerVarTest.java [hunk 11]
```diff
@@ -60,8 +60,24 @@ public static void setUp() throws Exception {
         server.startServer();
     }
 
+    @Test
+    public void testConfig() throws Exception {
+        runTest(server, APP_NAME + "/testConfig", "testServiceNameConfig");
+        runTest(server, APP_NAME + "/testConfig", "testSDKDisabledConfig");
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryConfigServerVarTest.java]: Why was this changed?
I assume it replaces the `@TestServlet` annotation, but I'm not sure why it was necessary to do so.

Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryConfigServerVarTest.java]: I moved this test from using `@TestServlet` to using `runTest` because `testWarningForConflictingValues()` is a log reading test. It felt nicer than mixing the two paradigms.

Comment (Azquelt) [reply dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryConfigServerVarTest.java]: On balance I prefer to keep a mix to reduce the boilerplate and reduce the risk of updating one place without updating the other when adding new tests to the servlet or refactoring, but it's not a big deal.

However, putting both tests in one method like this means you'll only get one result for both methods rather  than separate failures for each. If you're going to move them to the class like this, please keep them in separate methods.

Please also update the test servlet with a comment to point out that new tests added to the servlet also need to be added to the test class since that's not the norm.

Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/TelemetryConfigServerVarTest.java]: Since I'm changing them anyway and you asked I'll put it back to a mixed paradigm.

**Comment by benjamin-confino** [comment]
#build

**Review by Azquelt — APPROVED** [review]

**Review by helyarp — COMMENTED** [review]
Looks good!

**Comment by benjamin-confino** [comment]
[While pipeline 6a2f9d62-a7e2-4442-9e29-cc56562ea5a8 hasn't completed, it has gotten past the telemetry fats. ](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=6a2f9d62-a7e2-4442-9e29-cc56562ea5a8) has three failures, all intermittent, seen previously, and not telemetry related.
